Source/EasyNetQ.DI.Microsoft/ServiceCollectionAdapter.cs

```diff
@@ -1,78 +1,160 @@
+    }
+
+    /// <inheritdoc />
+    public IServiceRegister TryRegister(Type serviceType, Func<IServiceResolver, object> implementationFactory, Lifetime lifetime = Lifetime.Singleton, RegistrationCompareMode mode = RegistrationCompareMode.ServiceType)
+    {
+        if (serviceType == null)
+            throw new ArgumentNullException(nameof(serviceType));
+        if (implementationFactory == null)
+            throw new ArgumentNullException(nameof(implementationFactory));
+
+        var descriptor = new ServiceDescriptor(serviceType, PreserveFuncType(implementationFactory), ToLifetime(lifetime));
+        if (mode == RegistrationCompareMode.ServiceType)
+            serviceCollection.TryAdd(descriptor);
+        else if (mode == RegistrationCompareMode.ServiceTypeAndImplementationType)
+            serviceCollection.TryAddEnumerable(descriptor);
+        else
+            throw new ArgumentOutOfRangeException(nameof(mode));
+
+        return this;
     }
 
     /// <inheritdoc />
-    public IServiceRegister Register(
-        Type serviceType, Type implementingType, Lifetime lifetime = Lifetime.Singleton
-    )
+    public IServiceRegister TryRegister(Type serviceType, object implementationInstance, RegistrationCompareMode mode = RegistrationCompareMode.ServiceType)
     {
-        switch (lifetime)
+        if (serviceType == null)
+            throw new ArgumentNullException(nameof(serviceType));
+        if (implementationInstance == null)
+            throw new ArgumentNullException(nameof(implementationInstance));
+
+        var descriptor = new ServiceDescriptor(serviceType, implementationInstance);
+        if (mode == RegistrationCompareMode.ServiceType)
+            serviceCollection.TryAdd(descriptor);
+        else if (mode == RegistrationCompareMode.ServiceTypeAndImplementationType)
+            serviceCollection.TryAddEnumerable(descriptor);
+        else
+            throw new ArgumentOutOfRangeException(nameof(mode));
+
+        return this;
+    }
+
+    private static ServiceLifetime ToLifetime(Lifetime lifetime)
+        => lifetime switch
         {
-            case Lifetime.Transient:
-                serviceCollection.AddTransient(serviceType, implementingType);
-                return this;
-            case Lifetime.Singleton:
-                serviceCollection.AddSingleton(serviceType, implementingType);
-                return this;
-            default:
-                throw new ArgumentOutOfRangeException(nameof(lifetime), lifetime, null);
-        }
+            Lifetime.Singleton => ServiceLifetime.Singleton,
+            //Lifetime.Scoped => MSServiceLifetime.Scoped,
+            Lifetime.Transient => ServiceLifetime.Transient,
+            _ => throw new ArgumentOutOfRangeException(nameof(lifetime))
+        };
+
+    // Without this code, the type of return value will be object
+    private static Func<IServiceProvider, object> PreserveFuncType(Func<IServiceResolver, object> implementationFactory)
```

Comment:
@Shane32 Remember our work on GraphQL DI. We have no such problem there since we pass delegate direct to MS API. But here `Func<IServiceResolver, object>` is converted to `Func<IServiceProvider, object>` so the actual return type of initial `Func<,>` is lost.

Reply:
I remember us working on it, but I do not believe that statement is correct, for factory delegate registrations anyway.  The method signature is:

```cs
IServiceRegister Register(Type serviceType, Func<IServiceProvider, object> implementationFactory, ServiceLifetime serviceLifetime, bool replace = false)
```

We have no code to preserve the type as it is passed to MSDI.  Rather, MSDI has some special code to work their magic.

On a side note, I still do not think GraphQL DI interfaces should have support for TryAddEnumerable or replace semantics.  They are not needed or necessary to the project and are outside the scope of GraphQL.  The `IServiceRegister` interface should only serve the purposes of registering GraphQL services, and basic support for additional services as would typically be needed for options classes and similar.  TryAddEnumerable and replace semantics are advanced use cases not typically required for GraphQL add-on libraries.  Any advanced use case by a developer can register their services directly with the underlying dependency injection provider.

Reply:
I see.